cpp/include/cugraph/sampling_functions.hpp

```diff
@@ -240,6 +245,375 @@ biased_neighbor_sample(
+ * destination vertex, weight (optional) is the edge weight, edge_id (optional) identifies the edge
+ * id, edge_type (optional) identifies the edge type, hop identifies which hop the edge was
+ * encountered in.  The offsets array (optional) identifies the offset for each label.
+ *
+ * If @p starting_vertex_offsets is not specified then no organization is applied to the output, the
+ * offsets values in the return set will be std::nullopt.
+ *
+ * If @p starting_vertex_offsets is specified the offsets array will be populated. The offsets array
+ * in the return will be a CSR-style offsets array to identify the beginning of each label range in
+ * the output vectors.
+ *
+ * If @p label_to_output_comm_rank is specified then the data will be shuffled so that all entries
+ * for a particular label are returned on the specified rank.  This will result in the offsets array
+ * on other GPUs indicating that there are no entries for that label (`offsets[i] == offsets[i+1]`).
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam edge_t Type of edge identifiers. Needs to be an integral type.
+ * @tparam weight_t Type of edge weights. Needs to be a floating point type.
+ * @tparam edge_type_t Type of edge type. Needs to be an integral type.
+ * @tparam bias_t Type of bias. Needs to be an integral type.
+ * @tparam store_transposed Flag indicating whether sources (if false) or destinations (if
+ * true) are major indices
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param rng_state A pre-initialized raft::RngState object for generating random numbers
+ * @param graph_view Graph View object to generate NBR Sampling on.
+ * @param edge_weight_view Optional view object holding edge weights for @p graph_view.
+ * @param edge_id_view Optional view object holding edge ids for @p graph_view.
+ * @param edge_type_view Optional view object holding edge types for @p graph_view.
+ * @param starting_vertices Device span of starting vertex IDs for the sampling.
+ * In a multi-gpu context the starting vertices should be local to this GPU.
+ * @param starting_vertex_labels Optional device span of labels associated with each starting vertex
+ * for the sampling.
+ * @param label_to_output_comm_rank Optional device span identifying which rank should get each
+ * vertex label.  This should be the same on each rank.
+ * @param fan_out Host span defining branching out (fan-out) degree per source vertex for each
+ * level. The fanout value at hop x is given by the expression 'fanout[x*num_edge_types +
+ * edge_type_id]'
+ * @param num_edge_types Number of edge types where a value of 1 translates to homogeneous neighbor
+ * sample whereas a value greater than 1 translates to heterogeneous neighbor sample.
```

Comment:
What happens if the user passes 1 here? Throw an exception and asks to call the homogeneous version instead? Or internally call the homogeneous version? Or take the heterogeneous code path (which might be less efficient)?

Reply:
```
for (auto hop = 0; hop < num_hops; hop++) {
    for (auto edge_type_id = 0; edge_type_id < num_edge_types; edge_type_id++) {
```

Both homogenous and heterogeneous sampling use the same detail function where the latter attach a mask if `num_edge_types > 1` hence, if the user passes 1, no mask will leveraged. If you look at the inner loop above, it will only execute once, similar to the homogenous version which has the `num_edge_types` hardcoded to 1